controller/state.go

```diff
@@ -802,18 +818,25 @@ func (m *appStateManager) CompareAppState(app *v1alpha1.Application, project *v1
 
 // useDiffCache will determine if the diff should be calculated based
 // on the existing live state cache or not.
-func useDiffCache(noCache bool, manifestInfos []*apiclient.ManifestResponse, sources []v1alpha1.ApplicationSource, app *v1alpha1.Application, manifestRevisions []string, statusRefreshTimeout time.Duration, log *log.Entry) bool {
+func useDiffCache(noCache bool, manifestInfos []*apiclient.ManifestResponse, sources []v1alpha1.ApplicationSource, app *v1alpha1.Application, manifestRevisions []string, statusRefreshTimeout time.Duration, serverSideDiff bool, log *log.Entry) bool {
 
 	if noCache {
 		log.WithField("useDiffCache", "false").Debug("noCache is true")
 		return false
 	}
-	_, refreshRequested := app.IsRefreshRequested()
+	refreshType, refreshRequested := app.IsRefreshRequested()
 	if refreshRequested {
-		log.WithField("useDiffCache", "false").Debug("refreshRequested")
-		return false
+		switch refreshType {
+		case v1alpha1.RefreshTypeNormal:
+			log.WithField("useDiffCache", "false").Debug("normal refresh Requested")
+			return false
+		case v1alpha1.RefreshTypeHard:
+			log.WithField("useDiffCache", "false").Debug("hard refresh requested")
+			return false
+		}
```

Comment:
Want to add a quick comment about why we only discard the diff in these two cases?

Seems like the change is potentially not completely related to server side diff... if so, should we do it as a separate PR and maybe even cherry-pick it back?

Reply:
Those are the two only RefreshTypes existing today. This was added originally in this PR to still use cache even if a normal refresh is requested when ServerSide Diff is enabled. However I reverted that logic as it was causing inconsistent behaviour while switching from SSD to true/false.
Cleaned this code in my latest commit.